File: stack/largestRectangleArea.cpp
#include <iostream>
#include <stack>
#include <vector>

using std::stack;
using std::vector;

class Solution {
public:
    int largestRectangleArea(vector<int>& height) {
        stack<int> s;
        int result = 0;
        height.push_back(0);
        for (int i = 0; i < height.size(); ) {
            if (s.empty() || height[i] > height[s.top()])
                s.push(i++);
            else {
                int tmp = s.top();
                s.pop();
                result = std::max(result, height[tmp] * (s.empty() ? i : (i - s.top() - 1)));
            }
        }
        return result;
    }
};

int main()
{
    vector<int> s {2,1,5,6,6,6};
    Solution solu;
    std::cout << solu.largestRectangleArea(s) << "\n";
    return 0;
}

File: stack/largestRectangleArea.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        i = max_area = 0
        size = len(heights)
        while i < size:
            h = heights[i]
            if not stack or heights[stack[-1]] < h:
                stack.append(i)
                i += 1
            else:
                top = stack.pop()
                max_area = max(max_area,
                               heights[top] * (i - stack[-1] - 1 if stack else i))
        while stack:
            top = stack.pop()
            max_area = max(max_area,
                           heights[top] * (i - stack[-1] - 1 if stack else i))
        return max_area

    def largestRectangleArea1(self, heights: List[int]) -> int:
        stack = []
        heights.append(0)
        max_area = 0
        i = 0
        while (i < len(heights)):
            if (not stack) or (heights[i] > heights[stack[-1]]):
                stack.append(i)
                i += 1
            else:
                tmp = stack.pop()
                max_area = max(max_area,
                               heights[tmp] * ((i - stack[-1] - 1) if stack else i))
        return max_area

s = Solution()
h = [2,1,5,6,2,3]
print(s.largestRectangleArea1(h))
